### PR TITLE
Fixed decimal UTC time bug

### LINKED ISSUE #90: [Question] How to display decimal UTC offset?
I want to know if my UTC offset is UTC/GMT+5:30, how to display it in the URL? I tried it as 5.30, but all bars went even, means there was an error. Any Help would be Appreciated.

### PR DESCRIPTION
**Pull request by SauravDharwadkar**
fixed #90 
floor down value because chart uses hours only

### CODE REVIEW

**Comment by vn7n24fzkq** [comment]
Thanks for your work!!

I would like to merge this PR, but maybe we  can use `ground` here?

**Comment by vn7n24fzkq** [comment]
Never mind, I am just realize that we use `Math.floor` can avoid to turn 23.6 to 24, if we use `Math.round` will need to write some code to deal with some special cases,.

I think we can take this PR.


I will do some more review later, to check there are no other problems.

**Comment by SauravDharwadkar** [comment]
I thought that , utc time 0 , 0.3 , 0.45 in min where 0.6 = 1
So for avoid dealing with 0.5 nunber and 0.3 hour / 30 min i used floor.
We can add logic to round robin between hour or ceil if 0.45 .

**Comment by SauravDharwadkar** [comment]
Or we can simply check value after point , 0 / 3 / 45 and update utc accordingly and use floor for default to avoid incorrect utc format

**Comment by vn7n24fzkq** [comment]
> Or we can simply check value after point , 0 / 3 / 45 and update utc accordingly and use floor for default to avoid incorrect utc format 

I prefer this way, but throw an exception when the UTC number is invalid.
We can check if the value is invalid and throw an exception here to alert the caller the UTC format was incorrect.

And like you said, default to using floor.

**Comment by SauravDharwadkar** [comment]
When i looked chart  i realise we need to use `number.toFixed(2) // '6.00'` this or treat utc as string bcz in chart time should display as e.g. +5.30 or -5.30
But now its just shows 5.3 .
Both are same but first glance it look like something is wrong with utc time .

**Comment by vn7n24fzkq** [comment]
You are right, would you like to fix this in this PR?

**Comment by vn7n24fzkq** [comment]
Or we can create another issue for this if that need to modify many codes.

**Comment by SauravDharwadkar** [comment]
Seting up utc time on chart as well handling offset simple task , but I'm not sure how will you display warning or handle the wrong utc and dont want to lean whole code. I'll handle chart display and utc offset u do something about warning .

**Comment by vn7n24fzkq** [comment]
Sounds good to me, thanks!

**Comment by vn7n24fzkq** [comment]
By the way, about the warning, I will just throw an exception like this.

```js
if(!IsValidUTCTime(utcTime){
    throw Exception ("The UTC format is invalid, it should be ......")
}
```

**Comment by SauravDharwadkar** [comment]
const adjustOffset = function (offset: number, RoundRobin: {offset: number}): number {
    if (offset % 1 == 0) {
        return offset;
        // offset % 1 should be 0.3 or 0.7 but its js and it gives 0.29999 or -0.299999 thats why this frankenstein
    } else if ((offset % 1 > 0.29 && offset % 1 < 0.31) || (offset % 1 < -0.29 && offset % 1 > -0.31)) {
        // toggle up and down between hour
        RoundRobin.offset = (RoundRobin.offset + 1) % 2;
        return RoundRobin.offset === 0 ? Math.floor(offset) : Math.ceil(offset);
    } else if ((offset % 1 > 0.44 && offset % 1 < 0.46) || (offset % 1 < -0.44 && offset % 1 > -0.45)) {
        // distrubute 1 : 3 ratio for 0.45 utc time
        RoundRobin.offset = (RoundRobin.offset + 1) % 2;
        return RoundRobin.offset % 4 == 0 ? Math.floor(offset) : Math.ceil(offset);
    } else {
        // flood down , if utc is given right it will never be executed
        return Math.floor(offset);
    }
};

.
.
 let RoundRobin = {
        offset: 0
    };
.
.
 chartData[24 + adjustOffset(afterOffset, RoundRobin)] += 1;





or use simply this for 0.3 and 0.45
Math.ceil(afterOffset);

**Comment by SauravDharwadkar** [comment]
> By the way, about the warning, I will just throw an exception like this.
> 
> ```js
> if(!IsValidUTCTime(utcTime){
>     throw Exception ("The UTC format is invalid, it should be ......")
> }
> ```

in api IDK how it will display

**Comment by vn7n24fzkq** [comment]
> ```js
> const adjustOffset = function (offset: number, RoundRobin: {offset: number}): number {
>     if (offset % 1 == 0) {
>         return offset;
>         // offset % 1 should be 0.3 or 0.7 but its js and it gives 0.29999 or -0.299999 thats why this frankenstein
>     } else if ((offset % 1 > 0.29 && offset % 1 < 0.31) || (offset % 1 < -0.29 && offset % 1 > -0.31)) {
>         // toggle up and down between hour
>         RoundRobin.offset = (RoundRobin.offset + 1) % 2;
>         return RoundRobin.offset === 0 ? Math.floor(offset) : Math.ceil(offset);
>     } else if ((offset % 1 > 0.44 && offset % 1 < 0.46) || (offset % 1 < -0.44 && offset % 1 > -0.45)) {
>         // distrubute 1 : 3 ratio for 0.45 utc time
>         RoundRobin.offset = (RoundRobin.offset + 1) % 4;
>         return RoundRobin.offset % 4 == 0 ? Math.floor(offset) : Math.ceil(offset);
>     } else {
>         // flood down , if utc is given right it will never be executed
>         return Math.floor(offset);
>     }
> };
> 
> .
> .
>  let RoundRobin = {
>         offset: 0
>     };
> .
> .
>  chartData[24 + adjustOffset(afterOffset, RoundRobin)] += 1;
> 
> ```

I think we can take this code, and after that, we don't need to throw exceptions, tho.

**Comment by vn7n24fzkq** [comment]
> > By the way, about the warning, I will just throw an exception like this.
> > 
> > ```js
> > if(!IsValidUTCTime(utcTime){
> >     throw Exception ("The UTC format is invalid, it should be ......")
> > }
> > ```
> 
> in api IDK how it will display

We don't care about the error display here.

And  `Math.ceil(afterOffset);` is accepted too, if we use this way, I will take the UTC validation part.

**Comment by vn7n24fzkq** [comment]
@SauravDharwadkar Hi, thanks for your work. 

Are you still working on this?
Please let me know if there is any problem.

**Comment by SauravDharwadkar** [comment]
was busy .
i didn't test negative value bcz process.arg not work with negative value

File: src/cards/productive-time-card.ts [hunk 1]
```diff
@@ -39,10 +57,14 @@ const getProductiveTimeData = async function (username: string, utcOffset: numbe
     // process productiveTime
     const chartData = new Array(24);
     chartData.fill(0);
+    // eslint-disable-next-line prefer-const
```
Comment (vn7n24fzkq) [inline src/cards/productive-time-card.ts:60]: If possible, I think we should avoid disabling the `eslint` here. Could we do that? 🤔

File: src/cards/productive-time-card.ts [hunk 2]
```diff
@@ -39,10 +57,14 @@ const getProductiveTimeData = async function (username: string, utcOffset: numbe
...
+    let RoundRobin = {
```
Comment (vn7n24fzkq) [inline src/cards/productive-time-card.ts:61]: Use camel case, the first letter should be lowercase.

File: src/cards/productive-time-card.ts [hunk 3]
```diff
@@ -39,10 +57,14 @@ const getProductiveTimeData = async function (username: string, utcOffset: numbe
...
+        offset: 0
+    };
     for (const time of productiveTime.productiveDate) {
         const hour = new Date(time).getUTCHours(); // We use UTC+0 here
-        const afterOffset = Number(hour) + Number(utcOffset); // Add offset to hour
-        // covert afterOffset to 0-23
+        const afterOffset = adjustOffset(Number(hour) + Number(utcOffset), RoundRobin); // Add offset to hour
```
Comment (vn7n24fzkq) [inline src/cards/productive-time-card.ts:66]: Could you fix this?
I am just quickly google this one.
https://stackoverflow.com/questions/53516594/why-do-i-keep-getting-delete-cr-prettier-prettier

**Review by vn7n24fzkq — COMMENTED** [review]
Great!
I left some comments here, take a look, please.

**Comment by vn7n24fzkq** [comment]
Hi, @SauravDharwadkar.
Are you still working on this? 
Since this PR is almost done, if you don't mind I can take over the rest part and keep your commit in the commit history.

**Comment by SauravDharwadkar** [comment]
You can take over , sorry for late reply busy last couple of months

**Comment by vn7n24fzkq** [comment]
Solved in #123

**Pull request closed without merging**